File: server/src/providers/llm/localPersona.ts
import type { LlmProvider, LlmRequest } from '../types.js';

/**
 * Anahtarsiz calisan son care. Gercek zeka degil; karakterin ekranda yasadigini
 * dogrulamak icin kisa, karakterde replikler uretir. Gemini anahtari girilince
 * otomatik devre disi kalir.
 */
export class LocalPersonaProvider implements LlmProvider {
  readonly id = 'local';
  readonly label = 'Yerel karakter motoru (anahtarsiz)';

  private lastIndex = new Map<string, number>();

  async available(): Promise<boolean> {
    return true;
  }

  async *stream(req: LlmRequest): AsyncIterable<string> {
    const reply = this.compose(req.message);
    // Gercek streaming hissi: kelime kelime, insan temposunda.
    for (const word of reply.split(/(\s+)/)) {
      if (req.signal?.aborted) return;
      yield word;
      if (word.trim()) await sleep(18 + Math.random() * 26);
    }
  }

  private compose(message: string): string {
    const text = message.toLocaleLowerCase('tr');

    for (const rule of RULES) {
      if (rule.match.some((k) => text.includes(k))) {
        return this.pick(rule.key, rule.replies);
      }
    }
    if (text.trim().endsWith('?')) return this.pick('soru', QUESTION_REPLIES);
    return this.pick('genel', GENERIC_REPLIES);
  }

  private pick(key: string, list: string[]): string {
    if (list.length === 0) return '[IDLE] Hmm.';
    const prev = this.lastIndex.get(key) ?? -1;
    let index = Math.floor(Math.random() * list.length);
    if (list.length > 1 && index === prev) index = (index + 1) % list.length;
    this.lastIndex.set(key, index);
    return list[index] as string;
  }
}

const RULES: { key: string; match: string[]; replies: string[] }[] = [
  {
    key: 'selam',
    match: ['merhaba', 'selam', 'günaydın', 'gunaydin', 'iyi akşamlar', 'naber', 'nasılsın'],
    replies: [
      '[IDLE] Buradayım. Ne var?',
      '[AMUSED] Hoş geldin. Anlat bakalım.',
      '[IDLE] Hmm? Söyle.',
    ],
  },
  {
    key: 'abi',
    match: ['abi', 'abicim', 'abiciğim'],
    replies: [
      '[IDLE] Efendim. Dinliyorum.',
      '[AMUSED] Gel bakalım. Ne oldu?',
      '[SERIOUS] Buradayım. Söyle.',
    ],
  },
  {
    key: 'yardim',
    match: ['yardım', 'yardim', 'yapabilir misin', 'hallet', 'çöz', 'coz'],
    replies: [
      '[SERIOUS] Tamam, çekil. Ben hallederim.',
      '[SUSPICIOUS] Bir dakika. Önce ne yapmaya çalışıyorsun?',
      '[SERIOUS] Göster. Bakayım şuna.',
    ],
  },
  {
    key: 'oyun',
    match: ['oyun', 'maç', 'mac', 'kaybettim', 'yendim', 'rakip'],
    replies: [
      '[ANNOYED] Aynı şeyi kaçıncı kez yapıyorsun? Taktik değiştiriyoruz.',
      '[EXCITED] Tamam, şimdi oldu. Devam et.',
      '[AMUSED] Ben sana söylemiştim.',
    ],
  },
  {
    key: 'itiraz',
    match: ['satın al', 'satin al', 'para gönder', 'kredi', 'yatırım', 'yatirim'],
    replies: [
      '[ALERT] Yok. Bence hiç bulaşma.',
      '[SUSPICIOUS] Dur bakalım. Sen ciddi misin?',
    ],
  },
  {
    key: 'tesekkur',
    match: ['teşekkür', 'tesekkur', 'sağ ol', 'sag ol', 'eyvallah'],
    replies: ['[AMUSED] Ne demek. Devam.', '[IDLE] Tamam tamam.'],
  },
  {
    key: 'kimsin',
    match: ['kimsin', 'nesin', 'yapay zeka mısın', 'bot musun'],
    replies: [
      '[SERIOUS] Abinim. Bu kadarı yeter.',
      '[AMUSED] Buradayım işte. Başka?',
    ],
  },
];

const QUESTION_REPLIES = [
  '[SUSPICIOUS] Bir saniye. Tam olarak neyi soruyorsun?',
  '[SERIOUS] Bak şimdi. Önce şunu netleştirelim.',
  '[IDLE] Hmm. Biraz daha aç bakalım.',
];

const GENERIC_REPLIES = [
  '[IDLE] Tamam, dinliyorum. Devam et.',
  '[SUSPICIOUS] Dur. Nereye varmaya çalışıyorsun?',
  '[AMUSED] Haydaa. Bunu bir daha söyle.',
  '[SERIOUS] Anladım. Göster bakalım.',
];

function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: server_py/abi/providers/llm/local_persona.py
"""
Anahtarsiz calisan son care.

Gercek zeka degil; karakterin ekranda yasadigini dogrulamak icin kisa,
karakterde replikler uretir. Gemini anahtari girilince devre disi kalir.
"""

from __future__ import annotations

import asyncio
import random
import re
from typing import AsyncIterator

from .base import LlmRequest

_RULES: list[tuple[str, tuple[str, ...], tuple[str, ...]]] = [
    (
        "selam",
        ("merhaba", "selam", "günaydın", "gunaydin", "iyi akşamlar", "naber", "nasılsın"),
        ("[IDLE] Buradayım. Ne var?", "[AMUSED] Hoş geldin. Anlat bakalım.", "[IDLE] Hmm? Söyle."),
    ),
    (
        "abi",
        ("abi", "abicim", "abiciğim"),
        ("[IDLE] Efendim. Dinliyorum.", "[AMUSED] Gel bakalım. Ne oldu?", "[SERIOUS] Buradayım. Söyle."),
    ),
    (
        "yardim",
        ("yardım", "yardim", "yapabilir misin", "hallet", "çöz", "coz"),
        (
            "[SERIOUS] Tamam, çekil. Ben hallederim.",
            "[SUSPICIOUS] Bir dakika. Önce ne yapmaya çalışıyorsun?",
            "[SERIOUS] Göster. Bakayım şuna.",
        ),
    ),
    (
        "oyun",
        ("oyun", "maç", "mac", "kaybettim", "yendim", "rakip"),
        (
            "[ANNOYED] Aynı şeyi kaçıncı kez yapıyorsun? Taktik değiştiriyoruz.",
            "[EXCITED] Tamam, şimdi oldu. Devam et.",
            "[AMUSED] Ben sana söylemiştim.",
        ),
    ),
    (
        "itiraz",
        ("satın al", "satin al", "para gönder", "kredi", "yatırım", "yatirim"),
        ("[ALERT] Yok. Bence hiç bulaşma.", "[SUSPICIOUS] Dur bakalım. Sen ciddi misin?"),
    ),
    (
        "gorme",
        ("şuna bak", "suna bak", "ekrana bak", "görüyor musun", "goruyor musun"),
        (
            "[SUSPICIOUS] Göremiyorum. Gemini anahtarı olmadan gözlerim kapalı.",
            "[SERIOUS] Bakamıyorum şu an. Anahtarı bir tak da göreyim.",
        ),
    ),
    (
        "tesekkur",
        ("teşekkür", "tesekkur", "sağ ol", "sag ol", "eyvallah"),
        ("[AMUSED] Ne demek. Devam.", "[IDLE] Tamam tamam."),
    ),
    (
        "kimsin",
        ("kimsin", "nesin", "yapay zeka mısın", "bot musun"),
        ("[SERIOUS] Abinim. Bu kadarı yeter.", "[AMUSED] Buradayım işte. Başka?"),
    ),
]

_QUESTION = (
    "[SUSPICIOUS] Bir saniye. Tam olarak neyi soruyorsun?",
    "[SERIOUS] Bak şimdi. Önce şunu netleştirelim.",
    "[IDLE] Hmm. Biraz daha aç bakalım.",
)

_GENERIC = (
    "[IDLE] Tamam, dinliyorum. Devam et.",
    "[SUSPICIOUS] Dur. Nereye varmaya çalışıyorsun?",
    "[AMUSED] Haydaa. Bunu bir daha söyle.",
    "[SERIOUS] Anladım. Göster bakalım.",
)

_WORD = re.compile(r"\S+\s*")


class LocalPersonaProvider:
    id = "local"
    label = "Yerel karakter motoru (anahtarsiz)"
    vision = False

    def __init__(self) -> None:
        self._last: dict[str, int] = {}

    async def available(self) -> bool:
        return True

    async def stream(self, request: LlmRequest) -> AsyncIterator[str]:
        reply = self._compose(request.message)
        # Gercek streaming hissi: kelime kelime, insan temposunda.
        for word in _WORD.findall(reply):
            yield word
            await asyncio.sleep(0.018 + random.random() * 0.026)

    def _compose(self, message: str) -> str:
        text = message.casefold()
        for key, hints, replies in _RULES:
            if any(hint in text for hint in hints):
                return self._pick(key, replies)
        if text.strip().endswith("?"):
            return self._pick("soru", _QUESTION)
        return self._pick("genel", _GENERIC)

    def _pick(self, key: str, options: tuple[str, ...]) -> str:
        if not options:
            return "[IDLE] Hmm."
        index = random.randrange(len(options))
        # Ust uste ayni repligi verme.
        if len(options) > 1 and index == self._last.get(key, -1):
            index = (index + 1) % len(options)
        self._last[key] = index
        return options[index]
